File: src/lib/sessions.ts
import type { Session, ChatMessage } from "./types";

const sessions = new Map<string, Session>();

export function getOrCreateSession(sessionId: string): Session {
  if (!sessions.has(sessionId)) {
    sessions.set(sessionId, {
      id: sessionId,
      messages: [],
      ticket: {},
      stage: "collecting",
    });
  }
  return sessions.get(sessionId)!;
}

export function addMessage(sessionId: string, message: ChatMessage): void {
  const session = getOrCreateSession(sessionId);
  session.messages.push(message);
}

export function getMessages(sessionId: string): ChatMessage[] {
  return getOrCreateSession(sessionId).messages;
}

export function setStage(sessionId: string, stage: Session["stage"]): void {
  getOrCreateSession(sessionId).stage = stage;
}

export function setTicket(sessionId: string, ticket: Partial<Session["ticket"]>): void {
  getOrCreateSession(sessionId).ticket = ticket;
}

export function setTicketId(sessionId: string, ticketId: string): void {
  getOrCreateSession(sessionId).ticketId = ticketId;
}

export function getCollectedFields(sessionId: string): string[] {
  const t = getOrCreateSession(sessionId).ticket;
  return Object.entries(t)
    .filter(([_, v]) => v !== undefined && v !== "")
    .map(([k]) => k);
}

const REQUIRED_FIELDS = ["title", "description", "priority", "customer_email"];

export function allRequiredCollected(sessionId: string): boolean {
  const t = getOrCreateSession(sessionId).ticket;
  return REQUIRED_FIELDS.every((f) => t[f as keyof typeof t] !== undefined && String(t[f as keyof typeof t]).trim() !== "");
}

File: sessions.py
import uuid

sessions: dict[str, dict] = {}


def get_or_create(session_id: str) -> dict:
    if session_id not in sessions:
        sessions[session_id] = {
            "messages": [],
            "ticket": {},
            "stage": "collecting",
            "ticket_id": None,
        }
    return sessions[session_id]


def add_message(session_id: str, role: str, content: str):
    sess = get_or_create(session_id)
    sess["messages"].append({"role": role, "content": content})


def get_messages(session_id: str) -> list[dict]:
    return get_or_create(session_id)["messages"][-10:]


def set_stage(session_id: str, stage: str):
    sess = get_or_create(session_id)
    sess["stage"] = stage


def set_ticket(session_id: str, ticket: dict):
    sess = get_or_create(session_id)
    sess["ticket"] = ticket


def set_ticket_id(session_id: str, ticket_id: str):
    sess = get_or_create(session_id)
    sess["ticket_id"] = ticket_id


def get_collected_fields(session_id: str) -> list[str]:
    ticket = get_or_create(session_id)["ticket"]
    return [k for k, v in ticket.items() if v]


def get_stage(session_id: str) -> str:
    return get_or_create(session_id)["stage"]


def all_required_collected(session_id: str) -> bool:
    ticket = get_or_create(session_id)["ticket"]
    required = ["title", "description", "priority", "customer_email"]
    return all(ticket.get(k) for k in required)


ticket_store: dict[str, dict] = {}


def create_ticket(ticket: dict) -> str:
    ticket_id = f"TKT-{uuid.uuid4().hex[:8].upper()}"
    ticket_store[ticket_id] = ticket
    return ticket_id


def list_tickets() -> list[dict]:
    return [{"ticket_id": tid, **t} for tid, t in ticket_store.items()]
